test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala

```diff
@@ -63,32 +63,271 @@ class DoubleTest {
     }
   }
 
+  @inline
+  private def assertDoubleToString(expected: String, value: Double): Unit = {
+    /* The optimizer constant fold the string conversion of `<float literal>.toString()` directly within the JDK.
+     * However, because the rounding behavior of Double.toString differs between versions before and after JDK 19.
+     * For instance, in JDK 21, `Double.toString(double)` specifies "round to even" for boundary cases,
+     * whereas this specific boundary behavior seems not specified in JDK 17.
+     * https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Double.html#toString(double)
+     * https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#toString(double)
+     * To ensure stability, we are skipping comparisons against `value.toString()`, to avoid being depend on the
+     * specific JDK's string conversion implementation.
+     */
+    // assertEquals(expected, value.toString)
+    assertEquals(expected, toString(value))
+  }
+
+  @noinline
+  private def toString(value: Double): String = value.toString
+
   @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
-        executingInPureWebAssembly)
-    if (executingInJVM) {
-      assertEquals("0.0", 0.0.toString)
-      assertEquals("-0.0", (-0.0).toString)
+    if (executingInJVM || executingInPureWebAssembly) {
+      assertDoubleToString("0.0", 0.0)
+      assertDoubleToString("-0.0", -0.0)
     } else {
-      assertEquals("0", 0.0.toString)
-      assertEquals("0", (-0.0).toString)
+      assertDoubleToString("0", 0.0)
+      assertDoubleToString("0", -0.0)
     }
-    assertEquals("NaN", Double.NaN.toString)
-    assertEquals("Infinity", Double.PositiveInfinity.toString)
-    assertEquals("-Infinity", Double.NegativeInfinity.toString)
-    if (executingInJVM) {
-      assertEquals("5.0", 5.0.toString)
-      assertEquals("-5.0", (-5.0).toString)
+    assertDoubleToString("NaN", Double.NaN)
+    assertDoubleToString("Infinity", Double.PositiveInfinity)
+    assertDoubleToString("-Infinity", Double.NegativeInfinity)
+    if (executingInJVM || executingInPureWebAssembly) {
+      assertDoubleToString("5.0", 5.0)
+      assertDoubleToString("-5.0", -5.0)
     } else {
-      assertEquals("5", 5.0.toString)
-      assertEquals("-5", (-5.0).toString)
+      assertDoubleToString("5", 5.0)
+      assertDoubleToString("-5", -5.0)
     }
-    assertEquals("1.2", 1.2.toString)
+    assertDoubleToString("1.2", 1.2)
   }
 
-  @Test def toHexStringTest(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
+  @Test def toStringTest(): Unit = {
+    assumeTrue(
+        "The output of JS and JVM (depends on JDK version) differs slightly from the Ryu implementations.",
```

Comment:
Could we tweak Ryu's boundary conditions to always do the same as JS? JS has been well-specified since forever, and that's already what Scala.js-on-Wasm uses.